src/xcode/ENA/ENA/Source/Services/HealthCertificateValidation/HealthCertificateValidationService.swift

```diff
@@ -0,0 +1,555 @@
+protocol HealthCertificateValidationProviding {
+	func validate(
+		healthCertificate: HealthCertificate,
+		arrivalCountry: String,
+		validationClock: Date,
+		completion: @escaping (Result<HealthCertificateValidationReport, HealthCertificateValidationError>) -> Void
+	)
+}
+
+/*
+General flow:
+1. apply technical validation (expirationTimes)
+2. download new certificate value sets or load them from the cache
+3. download new acceptance rules or load them from the cache
+4. download new invalidation rules or load them from the cache
+5. assemble FilterParameter and ExternalParameter
+6. validate combined acceptance and invalidation rules
+7. return interpreted rules
+*/
+
+// swiftlint:disable:next type_body_length
+final class HealthCertificateValidationService: HealthCertificateValidationProviding {
+	
+	// MARK: - Init
+	
+	init(
+		store: Store,
+		client: Client,
+		vaccinationValueSetsProvider: VaccinationValueSetsProvider,
+		signatureVerifier: SignatureVerification = SignatureVerifier(),
+		validationRulesAccess: ValidationRulesAccessing = ValidationRulesAccess()
+	) {
+		self.store = store
+		self.client = client
+		self.vaccinationValueSetsProvider = vaccinationValueSetsProvider
+		self.signatureVerifier = signatureVerifier
+		self.validationRulesAccess = validationRulesAccess
+	}
+		
+	// MARK: - Protocol HealthCertificateValidationProviding
+	
+	func validate(
+		healthCertificate: HealthCertificate,
+		arrivalCountry: String,
+		validationClock: Date,
+		completion: @escaping (Result<HealthCertificateValidationReport, HealthCertificateValidationError>) -> Void
+	) {
+		
+		// 1. Apply technical validation
+		let expirationDate = Date(timeIntervalSince1970: TimeInterval(healthCertificate.cborWebTokenHeader.expirationTime))
+		
+		// NOTE: We expect here a HealthCertificate, which was already json schema validated at its creation time. So the JsonSchemaCheck will here always be true. So we only have to check for the expirationTime of the certificate.
+		guard expirationDate >= validationClock else {
+			Log.warning("Technical validation failed: expirationDate < validationClock. Expiration date: \(private: expirationDate), validationClock: \(private: validationClock)", log: .vaccination)
+			return completion(.failure(.TECHNICAL_VALIDATION_FAILED))
+		}
+		Log.info("Successfully passed technical validation. Proceed with updating value sets...", log: .vaccination)
+
+		proceedWithUpdatingValueSets(
+			healthCertificate: healthCertificate,
+			arrivalCountry: arrivalCountry,
+			validationClock: validationClock,
+			completion: completion
+		)
+	}
+	
+	// MARK: - Private
+	
+	private let store: Store
+	private let client: Client
+	private let vaccinationValueSetsProvider: VaccinationValueSetsProvider
+	private let signatureVerifier: SignatureVerification
+	private let validationRulesAccess: ValidationRulesAccessing
+	private var subscriptions = Set<AnyCancellable>()
+	
+	// MARK: - Flow
+	
+	private func proceedWithUpdatingValueSets(
```

Comment:
I understand what you mean, but i choosed this naming pattern to make it more visible, that we have here a straight flow, which the functions follow. 
Perhaps "continue" would be better? Otherwise the methods would be completely decouples, but like i said, i would not prefer this.

Reply:
"Otherwise the methods would be completely decouples"
Nice, I like decoupling ✌️

Reply:
i see, i see